Add BatchedTensorImpl

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38723 Add "batching rule" for torch.sum(tensor, dims)
* **#38424 Add BatchedTensorImpl**

On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add one or two batching rules
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

Differential Revision: [D21640917](https://our.internmc.facebook.com/intern/diff/D21640917)

zou3519 committed:
Add BatchedTensorImpl

On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

izdeby commented on c10/core/DispatchKey.h:
```diff
@@ -150,6 +150,7 @@ enum class DispatchKey : uint8_t {
   // There are a number of alternative modes which may want to handle before
   // autograd; for example, error checking, tracing, profiling or vmap.  They
   // go here.
+  Vmap,
```
nit: i would add 1-2 lines comment about the key it self.

cpuhrsch replied:
Why call it Vmap and not BatchedTensorKey or such?

izdeby commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,86 @@
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, 5>;
```
i would push '5' as a constant next to `kVmapMaxTensorDims`. Also, why 5?

zou3519 replied:
It's a hypothesis that users won't use more than 5 batch dimensions. In PyTorch we guarantee "best" performance on tensors of dimensionality less than or equal to 5 (TensorImpl's sizes/strides use `SmallVector<int, 5>`).

I'll toss the number into a constant as you suggested (y)

izdeby commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,86 @@
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
```
does "Ref" part in the name implies that it should return reference?

zou3519 replied:
BatchDimsRef is typedef'ed to be an `ArrayRef<BatchDim>`. So yes, it is a reference

izdeby commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,86 @@
+}
+
+inline BatchedTensorImpl* getBatched(Tensor tensor) {
+  return static_cast<BatchedTensorImpl*>(tensor.unsafeGetTensorImpl());
```
`unsafeGetTensorImpl`, why is unsafe? sounds... unsafe

zou3519 replied:
Unsafe because it returns the underlying TensorImpl. Users of our C++ API for instance should not be using it and the word "unsafe" deters them.

cpuhrsch replied:
Also, in my experience, it even gives you the impl if the tensor is constant. So you could modify a constant Tensor.

ezyang replied:
It might be more in line with the prevailing naming if this one is also called getBatched. In fact, this is indeed unsafe, because it is not valid to call this on a non-batched tensor.

ezyang replied:
In this PR, you use getBatched in conjunction with isBatched. So an even safer form would be maybeGetBatched() that returns a nullptr if the result is not a batched tensor impl.

zou3519 replied:
I renamed `getBatched` to `unsafeGetBatched` and I created a `maybeGetBatched` variant (thanks for the suggestions!).

I have some downstream code that relies on `unsafeGetBatched` so I'd like to keep it around but we should be able to delete it in the future

izdeby commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,86 @@
+}
+
+// Returns a bitset. If bit i is set, then that means dim i is a batchdim.
+inline std::bitset<kVmapMaxTensorDims> createIsBdimBitset(BatchDimsRef bdims) {
```
nit: i would expand the name to be `createIsBdimBitset` to be `createBatchedDimBitset` cause with the first name is not obvious what is going on.

zou3519 replied:
Sure

izdeby commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -0,0 +1,95 @@
+    }
+    non_bdim_count++;
+  }
+  TORCH_INTERNAL_ASSERT(false);
```
Can you, please, add a comment on why we shouldn't get here? Or add an error message that would be obvious.

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+
+// We assume this in a few other places in the codebase,
+// but there isn't a centralized definition.
+constexpr int64_t kVmapMaxTensorDims = 64;
```
Why do you choose this constraint / number in particular?

zou3519 replied:
If you create a tensor with more than 64 dims, a bunch of things in PyTorch will fail. Off of the top of my head:
- TensorIterator reduction operations will fail because they assume the input can have no more than 64 dims. (https://github.com/pytorch/pytorch/issues/23159)

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

ezyang commented:
This PR is very pleasantly small, thank you!

ezyang commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+// probably use <= 5 nested vmaps, but adjust this number as necessary.
+constexpr int64_t kBatchDimsStackSize = 5;
+
+struct BatchDim {
```
What is a batch dim? (Comment!)

ezyang commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
```
What's "bdims"? (Comment!)

ezyang commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+  BatchDimsRef bdims() const { return bdims_; }
+  const Tensor& value() const { return value_; };
+
+  int64_t actualDim(int64_t dim, bool wrap_dim = true) const;
```
What's actualDim? (comment!)

ezyang commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+
+  int64_t actualDim(int64_t dim, bool wrap_dim = true) const;
+
+  // Override a bunch of methods inherited from TensorImpl.
```
Noooooo more overrides :(   Is there a reason why BatchTensorImpl cannot just redundantly record the size/strides of its internal tensor?

zou3519 replied:
BatchedTensorImpl can handle sizes without overriding (it can set the inherited `sizes_` vector).

All of the overrides here are safety checks and internal asserts (e.g., I think that strides(), is_contiguous() should produce an error message but I'm not 100% sure). Another way to do the override is to add a branch for `isBatched(Tensor)` [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L209-L211) and [here](https://github.com/pytorch/pytorch/blob/176174a68ba2d36b9a5aaef0943421682ecc66d4/aten/src/ATen/templates/TensorBody.h#L194-L196).

I'm fine with deleting the overrides and punting modifying `strides()` and `is_contiguous()` to the future after I've done some benchmarking to make sure adding the branch doesn't affect performance too much.

ezyang replied:
Oh it looks like you're overriding them to raise errors. OK

ezyang commented on aten/src/ATen/BatchedTensorImpl.h:
```diff
@@ -0,0 +1,95 @@
+  TORCH_INTERNAL_ASSERT(!isBatched(tensor));
+  auto tensor_dim = tensor.dim();
+  TORCH_CHECK(
+      tensor_dim < kVmapMaxTensorDims,
```
`<=`?

ezyang commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -0,0 +1,104 @@
+
+// The following are publically exposed as methods of Tensor
+IntArrayRef BatchedTensorImpl::strides() const {
+  TORCH_CHECK(false, "NYI: Getting tensor strides inside of vmap");
```
Arguably you can give meaningful strides for a batch tensor. The idea is that you give the strides that, if correspondingly interpreted, would let you address the FIRST batch in the batched tensor impl (so the strides of the public dimensions, not the private one).

zou3519 replied:
I agree that there is an argument that you can give meaningful strides. The reason why I am hesitant to provide it is because when users ask for strides, they are doing one of the following:
- they're about to grab the .data_ptr() and do some low-level things. vmap in its current state can't really support that.
- they're using the stride information to do some "performance tricks"; e.g. if you know what the strides of two matrices A, B are, it might be more performant to multiply them in some order (AB vs B^T A^T). The way vmap reorders dimensions could invalidate said performance tricks

ezyang commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -0,0 +1,104 @@
+  // to asking "where does the 3rd (0-indexed) zero occur in the bitset?".
+  // The answer to that is index 5.
+  int64_t non_bdim_count = 0;
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
```
I feel like this should be implementable without a loop. Maybe https://stackoverflow.com/questions/7669057/find-nth-set-bit-in-an-int ? Don't know how portable the top liked answer is though.

ezyang replied:
Although, since you're creating the bitset on the fly, I guess there isn't actually asymptotic improvement here :o)

zou3519 replied:
I don't know what CPUs we support, but the instruction set (BMI2) used in that answer should be on Intel CPUs newer than 6/2013 and AMD from newer than 2015. I'll leave a comment in the code as a TODO to figure that out

ezyang commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -0,0 +1,104 @@
+Tensor addBatchDim(const Tensor& tensor, int64_t level, int64_t dim) {
+  if (!isBatched(tensor)) {
+    BatchDims bdims;
+    bdims.push_back({level, dim});
```
c++ nit: `bdims.emplace_back(level, dim)`

ezyang commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -0,0 +1,104 @@
+    return at::detail::make_tensor<BatchedTensorImpl>(tensor, std::move(bdims));
+  }
+  const auto* batched = getBatched(tensor);
+  BatchDims new_bdims = { batched->bdims().begin(), batched->bdims().end() };
```
C++ nit: It'd be better not to use braced initializer here, because it is easier to get confused with the initializer list form of this (in this case, it's unambiguous, but a reader might misinterpret as "take only the first and the last element of bdims", which is not the intent). So for example you could write instead: `BatchDims new_bdims(batched->bdims().begin(), batched->bdims().end())`. Maybe cc @smessmer for any further thoughts.

ezyang commented:
At a high level, this PR seems very sparsely tested. I understand a lot of this will be tested more rigorously later. Probably you'll get the most utility from more heavy unit testing on the fiddly public/private batch dimensions calculations; esp with consideration for edge cases here.

ezyang commented:
Looks promising! Looking forward to the next PRs in this set.

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add one or two batching rules
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add one or two batching rules
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Add BatchedTensorImpl"


On the way to adding initial vmap support, this is the implementation
for BatchedTensorImpl. Vmap (in future PRs) leverages Tensors backed by
BatchedTensorImpl to do its work.

For more context, here is an overview of the plan to add initial vmap support.
- [this PR] Add BatchedTensorImpl
- Add one or two batching rules
- Add vmap Python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- Miscellaneous handling for failure cases
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]
